inc/cappuccino/TMap.hpp

```diff
@@ -0,0 +1,565 @@
+     * Existing items that are not touched will TTL out at the previous uniform
+     * TTL time.
+     * @param ttl The new TTL value to apply to all new elements.
+     */
+    auto UpdateDefaultTtl(
+        std::chrono::milliseconds ttl) -> void;
+
+private:
+    struct Element;
+    struct TtlElement;
+
+    using KeyedIterator = typename std::unordered_map<KeyType, Element>::iterator;
+    using TtlIterator = typename std::deque<TtlElement>::iterator;
+    
+    struct Element {
+        /// The element's value.
+        ValueType m_value;
+        /// The iterator into the ttl data structure.
+        TtlIterator m_ttl_position;
+    };
+
+    struct TtlElement {
+        /// The point in time in  which this element's value expires.
+        std::chrono::steady_clock::time_point m_expire_time;
+        /// The iterator to the Key in the data structure.
+        KeyedIterator m_keyed_elements_position;
+    };
+
+     auto doInsertUpdate(
+        const KeyType& key,
+        ValueType&& value,
+        std::chrono::steady_clock::time_point expire_time,
+        Allow allow) -> bool;
+
+    auto doInsert(
+        const KeyType& key,
+        ValueType&& value,
+        std::chrono::steady_clock::time_point expire_time) -> void;
+
+    auto doUpdate(
+        KeyedIterator keyed_position,
+        ValueType&& value,
+        std::chrono::steady_clock::time_point expire_time) -> void;
+
+    auto doInsertTtlElement(TtlElement ttl_element) -> TtlIterator;
+
+    auto doDelete(
+        KeyedIterator keyed_elements_position) -> void;
+
+    auto doFind(
+        const KeyType& key) -> std::optional<ValueType>;
+
+    auto doPrune(
+        std::chrono::steady_clock::time_point now) -> size_t;
+
+    /// Thread lock for all mutations.
+    Lock<SyncType> m_lock;
+
+    /// The current number of elements in the cache.
+    size_t m_used_size { 0 };
+
+    /// The keyed lookup data structure, the value is the Element struct which includes the value and an iterator to the associated m_ttls TTlElement.
+    typename std::unordered_map<KeyType, Element> m_keyed_elements;
+
+    /// The ttl sorted deque.
+    typename std::deque<TtlElement> m_ttls;
+
+    /// The default TTL for every key value pair inserted into the cache.
+    std::chrono::milliseconds m_ttl;
+};
+} // namespace cappuccino
+
+namespace cappuccino
+{
+
+template <typename KeyType, typename ValueType, Sync SyncType>
+TMap<KeyType, ValueType, SyncType>::TMap(
+    std::chrono::milliseconds ttl,
+    float max_load_factor)
+{
+    m_ttl = ttl;
+    m_keyed_elements.max_load_factor(max_load_factor);
+}
+
+template <typename KeyType, typename ValueType, Sync SyncType>
+auto TMap<KeyType, ValueType, SyncType>::Insert(
+        const KeyType& key,
+        ValueType value,
+        std::chrono::milliseconds ttl,
+        Allow allow) -> bool
+{
+    std::lock_guard guard { this->m_lock };
+    const auto now = std::chrono::steady_clock::now();
```

Comment:
Wouldn't this result in strange behavior if we calculate now before potentially waiting for the lock? Our times could end up potentially very inaccurate depending on the lock duration. This was admittedly more of a design concern when working on the uniform time aware structures that previously just pushed back on the TtlElements without sorting.

Reply:
Good point, if taken before the lock things could expire while waiting for the lock, if taken after the lock then you have the freshest time stamp.  I think this is less of a concern on the other cache types since operations are constant time, the doPrune() on TMap and TSet are what really could cause an issue here.  Hmm!